tests/integrationv2/test_client_authentication.py

```diff
@@ -275,7 +275,16 @@ def test_tls_12_client_auth_downgrade(managed_process):
 
     # A s2n server built with OpenSSL1.0.2 and enabling client auth will downgrade the protocol to TLS1.2.
     # The downgrade occurs because openssl-1.0.2 doesn't support RSA-PSS signature scheme.
-    if "openssl-1.0.2" in get_flag(S2N_PROVIDER_VERSION):
+    #
+    # TLS 1.3 is disabled when s2n-tls is built with libressl and boringssl, so TLS 1.2 will be negotiated
+    # with these libcryptos as well. https://github.com/aws/s2n-tls/issues/3250
+    if any([
+        libcrypto_str in get_flag(S2N_PROVIDER_VERSION) for libcrypto_str in [
+            "openssl-1.0.2",
+            "libressl",
+            "boringssl"
+        ]
+    ]):
```

Comment:
This is pretty messy. Why did we implement this test as an integration test instead of a self-talk test? It's a pretty narrow edge case, and we have much more control in our unit tests. We also have methods to check for rsa-pss / tls1.3 support.

Reply:
I couldn't think of an easy way to write this as a self-talk test. The client needs to send a TLS 1.3 client hello, but the libcrypto can't support RSA-PSS for the server to downgrade, but this causes the client to downgrade and send a TLS 1.2 client hello. It seems like you would have to construct a TLS 1.3 client hello manually or save a TLS 1.3 client hello and replay it? Or maybe there's a better way?

Reply:
Ah, I see the problem. Yeah, that'd probably be a very messy unit test, since all those "is_supported" checks are set at compile time.